File: worker/agents/voice.ts
import type Redis from "ioredis";
import { makeBus } from "../lib/bus";
import { recordSignal } from "../loops/l1";

export interface NarrateJob {
  uid: string;
  jobId: string;
  text?: string;
  articleId?: string;
  topic?: string;
}

/**
 * Voice agent: resolves the script to narrate and emits it for the client to speak
 * via Deepgram (/api/tts). Actual audio synthesis stays in the browser; this agent
 * owns selecting/shaping the narration text and recording the listen signal.
 */
export async function runNarrate(redis: Redis, job: NarrateJob): Promise<void> {
  const bus = makeBus(redis);
  let text = (job.text ?? "").trim();

  if (!text && job.articleId) {
    try {
      const raw = await redis.get(`article:${job.articleId}`);
      if (raw) {
        const a = JSON.parse(raw);
        text = String(a.summary || a.text || "").trim();
      }
    } catch {
      /* ignore */
    }
  }

  if (!text) {
    await bus.publish(job.uid, { jobId: job.jobId, type: "error", status: "error", error: "Nothing to narrate" });
    return;
  }

  await recordSignal(redis, job.uid, { kind: "listen", topic: job.topic });
  await bus.publish(job.uid, {
    jobId: job.jobId,
    type: "done",
    status: "ready",
    result: { text: text.slice(0, 1500) },
  });
}

File: worker_py/agents/voice.py
"""Voice agent. Port of ``worker/agents/voice.ts``.

Resolves the script to narrate and emits it for the client to speak via
Deepgram (/api/tts). Audio synthesis stays in the browser; this agent owns
selecting/shaping the narration text and recording the listen signal.
"""

import json
from typing import Dict

from redis.asyncio import Redis

from ..bus import publish as bus_publish
from ..loops.l1 import record_signal


async def run_narrate(redis: Redis, job: Dict[str, str]) -> None:
    uid, job_id = job["uid"], job["jobId"]
    text = (job.get("text") or "").strip()

    if not text and job.get("articleId"):
        try:
            raw = await redis.get(f"article:{job['articleId']}")
            if raw:
                a = json.loads(raw)
                text = str(a.get("summary") or a.get("text") or "").strip()
        except Exception:  # noqa: BLE001
            pass

    if not text:
        await bus_publish(redis, uid, {"jobId": job_id, "type": "error", "status": "error", "error": "Nothing to narrate"})
        return

    await record_signal(redis, uid, {"kind": "listen", "topic": job.get("topic")})
    await bus_publish(redis, uid, {"jobId": job_id, "type": "done", "status": "ready", "result": {"text": text[:1500]}})
